### PR TITLE
[Model] Enable batched video encoding for NanoNemotron-VL (RADIO ViT)

### PR DESCRIPTION
**Pull request by askliar**
## Summary

- **Removes the `requires_sequential_video_encoding` hack** from `GPUModelRunner` and the NanoNemotron model class. Videos are no longer processed one-at-a-time in a Python loop; instead, multiple videos are batched together through the ViT encoder in a single forward pass.
- **Adds two video encoding paths in NanoNemotron**: a **same-resolution fast path** that micro-batches up to ~128 frames at a time across multiple videos, and a **varying-resolution dynamic path** that pre-patches frames into the dynamic-resolution ViT format with per-tubelet positional encoding.
- **Refactors `ViTPatchGenerator.forward_video`** to accept `num_frames: list[int]` (multiple videos) instead of a single video tensor, producing all tubelets in one batched call.
- **Adds `forward_video_dynamic`** to `ViTPatchGenerator` for handling videos with heterogeneous spatial resolutions via temporal grouping and dynamic-resolution positional encoding.
- **Sorts multimodal items by modality** in `GPUModelRunner._execute_mm_encoder` so that `group_and_batch_mm_kwargs` sees consecutive same-modality items, enabling cross-request batching.

## Motivation

The previous implementation required sequential (one-video-at-a-time) encoding for NanoNemotron-VL's dynamic-resolution video pipeline because it did not support batch size > 1. This was a significant throughput bottleneck, especially for concurrent video requests. This PR lifts that restriction by:

1. Teaching the RADIO ViT patch generator to process multiple videos (with potentially different frame counts) in a single batched call.
2. Splitting the video embedding logic into a fast path (same H/W across videos) and a dynamic path (varying H/W), each capable of batching.
3. Removing the per-video loop and `requires_sequential_video_encoding` flag from the GPU model runner.

## Changes

| File | What changed |
|---|---|
| `nano_nemotron_vl.py` | Removed `requires_sequential_video_encoding` flag. Split `_extract_video_embeddings_temporal` into `_fast` (micro-batch windows, same-res) and `_dynamic` (pre-patch, varying-res) variants. Updated `_get_video_pixel_inputs` to preserve list form when video frames have different spatial sizes. |
| `radio.py` | Added `_apply_temporal_grouping` for dynamic-res tubelet grouping. Added `forward_video_dynamic` entry point. Refactored `forward_video` to accept `num_frames: list[int]` and batch across videos. Updated `RadioInternVisionModel.forward` with three-way dispatch (dynamic-res video / same-res video / images). Wired `_tubelet_imgs_sizes` into `RadioModel` for correct feature extraction. |
| `gpu_model_runner.py` | Removed `requires_sequential_video_encoding` flag and the sequential per-video fallback loop. Added modality-based sorting before batching so same-modality items are grouped across requests. Simplified encoder execution to always use the batched path. |

## Test plan

- [x] Serve model and verify single image, single video, and mixed image+video requests produce correct outputs.
- [x] Test batched inference with multiple concurrent video requests of the same resolution (exercises the fast path).
- [x] Test batched inference with multiple concurrent video requests of different resolutions (exercises the dynamic path).
- [x] Verify that `GPUModelRunner` no longer falls back to sequential encoding for video modality.
- [x] Run existing NanoNemotron-VL tests to check for regressions.

### CODE REVIEW

**Comment by netanel-haber** [comment]
Hello, please add assertions to the description: 
* Evals you've run before and after (that would trigger batches) showing the accuracy isn't affected. [There is no public model, so you can just write: "Ran DailyOmni and got equivalent results" 
* Assert you've done basic profiling that shows BS=1 throughput isn't hurt [you can use `vllm bench`], and that BS>1 is equivalent or better, I guess

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 1]
```diff
@@ -1132,22 +1129,178 @@ def _extract_video_embeddings_temporal(
...
+            pixel_values,
+            num_frames_per_video,
+            hidden_size,
+            is_list,
+            T,
+            patch_size,
+        )
+
+    def _extract_video_embeddings_temporal_fast(
+        self,
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py:1396]: Please make these arguments, or at least after `pixel_values`, keyword-only
```suggestion
        self, *,
```

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 2]
```diff
@@ -1132,22 +1129,178 @@ def _extract_video_embeddings_temporal(
...
+        return self._extract_video_embeddings_temporal_fast(
+            pixel_values,
+            num_frames_per_video,
+            hidden_size,
+            is_list,
+            T,
+            patch_size,
+        )
+
+    def _extract_video_embeddings_temporal_fast(
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: I think you can call this e.g., since fast doesn't communicate what the function does/needs, really

```suggestion
    def _extract_video_embeddings_temporal_same_resolution((
```

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 3]
```diff
@@ -1132,22 +1129,178 @@ def _extract_video_embeddings_temporal(
...
+        self,
+        pixel_values,
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Type annotation for `pixel_values`, please

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 4]
```diff
@@ -1132,22 +1129,178 @@ def _extract_video_embeddings_temporal(
...
+        num_frames_per_video: list[int],
+        hidden_size: int,
+        is_list: bool,
+        T: int,
+        patch_size: int,
+    ) -> tuple[torch.Tensor, ...]:
+        """Same-resolution fast path with micro-batch windows."""
+        micro_batch_size = 128 - (128 % T)
+
+        all_frames = torch.cat(pixel_values, dim=0) if is_list else pixel_values
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: I think `is_list` is guaranteed to be `False` for same resolution, right? If so, I wouldn't pass it here and do this logic [maybe just assert before calling the function, so you can safely have `tensor` be the type of pixel_values].

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -2793,6 +2792,19 @@ def _execute_mm_encoder(
...
 
+        # Sort by modality so that group_and_batch_mm_kwargs sees consecutive
+        # same-modality items and can batch them across requests.
+        # Example: [video_req1, audio_req1, video_req2, audio_req2]
+        #       -> [audio_req1, audio_req2, video_req1, video_req2]
+        # This resolves the FIXME above; mm_hashes is kept in sync so that
+        # the cache-by-hash logic below (zip(mm_hashes, encoder_outputs))
+        # remains correct.
+        _perm = sorted(range(len(mm_kwargs)), key=lambda i: mm_kwargs[i][0])
+        if _perm != list(range(len(_perm))):
```
Comment (netanel-haber) [inline vllm/v1/worker/gpu_model_runner.py]: Probably remove this line

File: vllm/v1/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -2803,72 +2815,17 @@ def _execute_mm_encoder(
...
-            # output vision tokens count should be considered)
-            # dynamic res video for nemotron temporarily uses this hack via
-            # requires_sequential_video_encoding
-            # because it doesn't yet support video batching.
-            # TODO(ywang96): Fix memory profiling to take EVS into account and
-            # remove this hack.
-            if (
-                (
-                    self.is_multimodal_pruning_enabled
-                    or self.requires_sequential_video_encoding
```
Comment (netanel-haber) [inline vllm/v1/worker/gpu_model_runner.py:2890]: Bring back this path, but have it branch only on *non* nemotron-nano-vl models (qwen etc.)

File: vllm/v1/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -2803,72 +2815,17 @@ def _execute_mm_encoder(
...
-                # 1. A tensor of shape (num_items, feature_size, hidden_size)
-                # in case feature_size is fixed across all multimodal items.
-                # 2. A list or tuple (length: num_items) of tensors,
-                # each of shape (feature_size, hidden_size) in case the feature
-                # size is dynamic depending on the input multimodal items.
-
-                with self.timed_encoder_operation(
-                    should_time, mm_lora_refs, current_item_idx, num_items
-                ):
-                    cudagraph_output = None
```
Comment (netanel-haber) [inline vllm/v1/worker/gpu_model_runner.py:2928]: Return this code

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 8]
```diff
@@ -1132,22 +1129,212 @@ def _extract_video_embeddings_temporal(
...
+                cur_size = 0
+            cur_segs.append((global_offset, nf))
+            cur_size += nf
+            global_offset += nf
+
+        if cur_segs:
+            windows.append(cur_segs)
+
+        results: list[torch.Tensor] = []
+        for window in windows:
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: This windowing scheme doesn't preserve the previous behavior of calling `extract_feature`, where at most 128 frames would be processed at once, to prevent OOMs

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 9]
```diff
@@ -1132,22 +1129,212 @@ def _extract_video_embeddings_temporal(
...
+        T: int,
+        patch_size: int,
+    ) -> tuple[torch.Tensor, ...]:
+        """Varying-resolution path: pre-patch frames to dynamic
+        format and process with per-tubelet positional encoding.
+
+        Videos are grouped into micro-batch windows by total patch count
+        to cap peak memory, mirroring the fast path's frame-based windowing.
+        """
+        # 128 frames at base 256 patches/frame (224×224, P=14).
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Please be extra verbose with this comment/docstring - explain that this 32768 magic number originates in 128 micro batch size, etc.

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 10]
```diff
@@ -1280,7 +1467,20 @@ def _parse_and_validate_video_input(
                 pixel_values_flat_video = list(pixel_values_flat_video)
 
             if not torch.is_tensor(pixel_values_flat_video):
-                pixel_values_flat_video = torch.cat(pixel_values_flat_video, dim=0)
+                if all(
+                    t.shape[-2:] == pixel_values_flat_video[0].shape[-2:]
+                    for t in pixel_values_flat_video
+                ):
+                    pixel_values_flat_video = torch.cat(pixel_values_flat_video, dim=0)
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: is this path still necessary? Isn't it possible that this path already filters for multi-shape, and therefore only the else should be relevant?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 11]
```diff
@@ -1280,7 +1467,20 @@ def _parse_and_validate_video_input(
                 pixel_values_flat_video = list(pixel_values_flat_video)
 
             if not torch.is_tensor(pixel_values_flat_video):
-                pixel_values_flat_video = torch.cat(pixel_values_flat_video, dim=0)
+                if all(
+                    t.shape[-2:] == pixel_values_flat_video[0].shape[-2:]
```
Comment (netanel-haber) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Check the width as well. Also: Maybe create a small named helper in-file for "is this multi shapen"?

File: vllm/model_executor/models/radio.py [hunk 12]
```diff
@@ -738,14 +870,15 @@ def forward(
         pixel_embeds: torch.Tensor | None = None,
         *,
         imgs_sizes: list[tuple[int, int]] | None = None,
-        num_frames: int | None = None,
+        num_frames: int | list[int] | None = None,
```
Comment (netanel-haber) [inline vllm/model_executor/models/radio.py]: Can this still be an `int`?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 13]
```diff
@@ -1234,22 +1231,212 @@ def _extract_video_embeddings_temporal(
...
+            cur_window.append(i)
+            cur_patches += n_patches
+
+        if cur_window:
+            windows.append(cur_window)
+
+        ds = self.downsample_ratio
+        results: list[torch.Tensor] = []
+
+        for window in windows:
```
Comment (askliar) [inline vllm/model_executor/models/nano_nemotron_vl.py]: rename window -> microbatch

**Comment by askliar** [comment]
@netanel-haber @collinmccarthy Ran batching experiments with DailyOmni and QBench Video MCQ. Speed improves significantly with both while accuracy stays the same or even slightly higher.

This PR is ready for review.
